File: src/tafqeetj/converters/IntegerToWordsConverter.java
package tafqeetj.converters;

import tafqeetj.exceptions.NumberOutOfRangeException;
import tafqeetj.numbers.NumberSign;
import tafqeetj.numbers.ThreeDigitsNumber;

import java.math.BigInteger;
import java.util.Map;

class IntegerToWordsConverter {
    private final NumberToThreeDigitsNumberGroupsDivider divider = new NumberToThreeDigitsNumberGroupsDivider();
    private static IntegerToWordsConverter instance;

    private IntegerToWordsConverter() {
    }

    public synchronized static IntegerToWordsConverter getInstance() {
        if (instance == null)
            instance = new IntegerToWordsConverter();
        return instance;
    }

    public IntegerInWords convert(int number) {
        return convert((long) number);
    }

    public IntegerInWords convert(byte number) {
        return convert((long) number);
    }

    public IntegerInWords convert(short number) {
        return convert((long) number);
    }

    public IntegerInWords convert(BigInteger number) {
        if (TafqeetRangeChecker.isOutOfRange(number))
            throw new NumberOutOfRangeException();
        return convert(number.longValue());
    }

    public IntegerInWords convert(long number) {
        if (TafqeetRangeChecker.isOutOfRange(number))
            throw new NumberOutOfRangeException();
        if (number == 0)
            return IntegerInWords.ZERO;
        IntegerInWords convertedNumber = convertNumber(divider.divide(number));
        if (number < 0)
            convertedNumber.setSign(NumberSign.NEGATIVE);
        return convertedNumber;
    }

    private IntegerInWords convertNumber(Map<String, ThreeDigitsNumber> threeDigitsNumberMap) {
        IntegerInWords integerInWords = new IntegerInWords();
        integerInWords.setTrillions(convertThreeDigitsNumberIfPresent("trillions", threeDigitsNumberMap));
        integerInWords.setBillions(convertThreeDigitsNumberIfPresent("billions", threeDigitsNumberMap));
        integerInWords.setMillions(convertThreeDigitsNumberIfPresent("millions", threeDigitsNumberMap));
        integerInWords.setThousands(convertThreeDigitsNumberIfPresent("thousands", threeDigitsNumberMap));
        integerInWords.setBasicNumberGroup(convertBasicNumberIfPresent(threeDigitsNumberMap));
        return integerInWords;
    }

    private String convertBasicNumberIfPresent(Map<String, ThreeDigitsNumber> threeDigitsNumberMap) {
        if (threeDigitsNumberMap.containsKey("basicNumberGroup")) {
            ThreeDigitsNumber number = threeDigitsNumberMap.get("basicNumberGroup");
            return number.getSign()
                    .getBasicNumberConverter()
                    .convert(number);
        }
        return "";
    }

    private String convertThreeDigitsNumberIfPresent(String numberGroupName,
                                                     Map<String, ThreeDigitsNumber> threeDigitsNumberMap) {
        if (threeDigitsNumberMap.containsKey(numberGroupName)) {
            ThreeDigitsNumber number = threeDigitsNumberMap.get(numberGroupName);
            return number.getSign()
                    .getQuantitiveNumberConverterFactory()
                    .getConverter(numberGroupName)
                    .convert(number);
        }
        return "";
    }
}

File: src/main/kotlin/tech/moussa/tafqeetk/converters/IntegerToWordsConverter.kt
package tech.moussa.tafqeetk.converters

import tech.moussa.tafqeetk.exceptions.NumberOutOfRangeException
import tech.moussa.tafqeetk.numbers.NumberSign
import tech.moussa.tafqeetk.numbers.ThreeDigitsNumber
import java.math.BigInteger

class IntegerToWordsConverter private constructor() {
    private val divider = NumberToThreeDigitsNumberGroupsDivider()

    fun convert(number: Int): IntegerInWords {
        return this.convert(number.toLong())
    }

    fun convert(number: Byte): IntegerInWords {
        return this.convert(number.toLong())
    }

    fun convert(number: Short): IntegerInWords {
        return this.convert(number.toLong())
    }

    fun convert(number: BigInteger): IntegerInWords {
        if (TafqeetRangeChecker.isOutOfRange(number)) {
            throw NumberOutOfRangeException(number)
        }
        return this.convert(number.toLong())
    }

    fun convert(number: Long): IntegerInWords {
        if (TafqeetRangeChecker.isOutOfRange(number)) {
            throw NumberOutOfRangeException(number)
        }
        if (number == 0L) {
            return IntegerInWords.ZERO
        }
        val convertedNumber = this.convertNumber(this.divider.divide(number))
        if (number < 0) {
            convertedNumber.setSign(NumberSign.NEGATIVE)
        }
        return convertedNumber
    }

    private fun convertNumber(threeDigitsNumberMap: Map<String, ThreeDigitsNumber>): IntegerInWords {
        val integerInWords = IntegerInWords()
        integerInWords.setTrillions(this.convertThreeDigitsNumberIfPresent("trillions", threeDigitsNumberMap))
        integerInWords.setBillions(this.convertThreeDigitsNumberIfPresent("billions", threeDigitsNumberMap))
        integerInWords.setMillions(this.convertThreeDigitsNumberIfPresent("millions", threeDigitsNumberMap))
        integerInWords.setThousands(this.convertThreeDigitsNumberIfPresent("thousands", threeDigitsNumberMap))
        integerInWords.setBasicNumberGroup(this.convertBasicNumberIfPresent(threeDigitsNumberMap))
        return integerInWords
    }

    private fun convertBasicNumberIfPresent(threeDigitsNumberMap: Map<String, ThreeDigitsNumber>): String {
        if (threeDigitsNumberMap.containsKey("basicNumberGroup")) {
            val number: ThreeDigitsNumber = threeDigitsNumberMap.getOrElse("basicNumberGroup") {
                return ""
            }
            return number.getSign().numberConverter.convert(number)
        }
        return ""
    }

    private fun convertThreeDigitsNumberIfPresent(numberGroupName: String, threeDigitsNumberMap: Map<String, ThreeDigitsNumber>): String {
        if (threeDigitsNumberMap.containsKey(numberGroupName)) {
            val number: ThreeDigitsNumber = threeDigitsNumberMap.getOrElse(numberGroupName) {
                return ""
            }
            return number.getSign().value.getConverter(numberGroupName).convert(number)
        }
        return ""
    }

    companion object {
        private var instance: IntegerToWordsConverter? = null

        @Synchronized
        fun getInstance(): IntegerToWordsConverter {
            if (IntegerToWordsConverter.instance == null) {
                IntegerToWordsConverter.instance = IntegerToWordsConverter()
            }
            return IntegerToWordsConverter.instance!!
        }
    }
}